src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,149 @@
+            Workspace workspace,
+            IDocumentNavigationService documentNavigationService)
+        {
+            _languageServiceBroker = languageServiceBroker;
+            _visualStudioCodeWindowInfoService = visualStudioCodeWindowInfoService;
+            _textViewEventSource = textViewEventSource;
+            _workspace = workspace;
+            _navigationService = documentNavigationService;
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            // initialize public properties
+            _sortOption = SortOption.Location;
+            _documentSymbolViewModelItems = new ObservableCollection<DocumentSymbolDataViewModel>();
+
+            // event queues for updating view model state
+            _expandCollapseQueue = new AsyncBatchingWorkQueue<ExpansionOption>(
+                DelayTimeSpan.NearImmediate,
+                ExpandOrCollapseItemsAsync,
+                asyncListener,
+                CancellationToken);
+
+            _selectTreeNodeQueue = new AsyncBatchingWorkQueue<CaretPosition>(
+                DelayTimeSpan.Short,
+                SelectTreeNodeAsync,
+                asyncListener,
+                CancellationToken);
+
+            _navigationQueue = new AsyncBatchingWorkQueue<TextSpan>(
+                DelayTimeSpan.NearImmediate,
+                NavigateToTextSpanAsync,
+                asyncListener,
+                CancellationToken);
+
+            _filterQueue = new AsyncBatchingWorkQueue<string>(
+                DelayTimeSpan.Short,
+                FilterTreeAsync,
+                asyncListener,
+                CancellationToken);
+
+            // work queues for refreshing LSP data
+            _documentSymbolRequestInfoQueue = new AsyncBatchingResultQueue<DocumentSymbolRequestInfo?>(
+                DelayTimeSpan.Short,
+                GetDocumentSymbolRequestInfoAsync,
+                asyncListener,
+                CancellationToken);
+
+            _documentSymbolQueue = new AsyncBatchingWorkQueue<DocumentSymbolRequestInfo, DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                GetDocumentSymbolAsync,
+                EqualityComparer<DocumentSymbolRequestInfo>.Default,
+                asyncListener,
+                CancellationToken);
+
+            _textViewEventSource.Changed += OnEventSourceChanged;
+            _textViewEventSource.Connect();
+
+            // queue initial model update
+            var service = _visualStudioCodeWindowInfoService.GetServiceAndThrowIfNotOnUIThread();
+            var info = service.GetDocumentSymbolRequestInfo();
+            Assumes.NotNull(info);
+            _documentSymbolQueue.AddWork(info, cancelExistingWork: true);
+        }
+
+        private SortOption _sortOption;
+        public SortOption SortOption
+        {
+            get => _sortOption;
+            set => SetProperty(ref _sortOption, value);
+        }
+
+        private readonly SemaphoreSlim _guard = new(1);
```

Comment:
move to top.

Reply:
i don't actually see what this is used for.  needs docs explaining what it is guarding.